Feature/addition of image studio sdk

bennealon opened the pull request:
Initial addition of image-studi-sdk
Interface to launch the studio, and wait for a response.
Handles message communication between the studio and any third party libraries that integrate with this SDK.
Adds basic Storybook implementation, predominantly to showcase and test the implementation.

bennealon committed:
MAN-1389 Addition of ImageStudio SDK
Adds Storybook for ImageStudio

bennealon committed:
Adds in ImageStudio Storybook, allows launch and test of the basic SDK
Updates packages to latest
Adds missing linter packages, updates config

bennealon committed:
linter fix

bennealon committed:
Updates gql-client library, and uses the new beta file provider
Updates SDK response to contain a `reason` to improve the interface between user and SDK
Adds two automated storybook tests to check both return paths. Could improve these in future by mocking the temporary file service response.

bennealon committed:
splitting into seperate files
minor readability improvements
Story updates to prevent rendering a faux form for test story's

bennealon committed:
Updates SDK to send image name

bennealon committed:
Updates SDK to match latest Image-Studio

bennealon committed:
MAN-1453 - image-studio launches via new tab
SDK launch improvements
Aligns error flow to that of content-studio-sdk

bennealon committed:
reduces code complexity
adds back in image-studio `focus` message

bennealon committed:
removes redundant interface

rs-amp commented on src/types/Message.types.ts:
```diff
@@ -0,0 +1,21 @@
+export interface SDKEvent {
```
Indentation is odd in this file. (every line after the first has +2 spaces?)

rs-amp commented on src/types/ImageStudio.types.ts:
```diff
@@ -0,0 +1,18 @@
+/**
+ * USed to capture the Promise resolution reason
```
```suggestion
 * Used to capture the Promise resolution reason
```

rs-amp commented on src/types/ImageStudio.types.ts:
```diff
@@ -0,0 +1,18 @@
+}
+
+/**
+ * ImageStudio.launch response
```
```suggestion
 * Image Studio launch response
```

rs-amp commented on src/errors/ApplicationBlockedError.ts:
```diff
@@ -0,0 +1,6 @@
+export class ApplicationBlockedError extends Error {
+    constructor() {
```
More indentation weirdness.

rs-amp commented on src/utils/Helpers.ts:
```diff
@@ -0,0 +1,24 @@
+export const getFileExtensionFromMimeType = (
+    mimeType: string | null,
```
Indentation is odd here too.

bennealon committed:
chore: function renames

rs-amp reviewed (commented):
I'm getting this with the test image in storybook:

```
Image size too large
This image exceeds Image Studio's size limit of 16MP (megapixels). Please rescale before uploading, or choose a different image.
```

Is there a guard against calling `launch` again while an instance is already running? Because it can only handle one result promise per instance, maybe it should throw if you try to stack instances.

amplience-dling commented on package.json:
```diff
@@ -4,33 +4,51 @@
-  "author": "",  
+  "author": "",
   "devDependencies": {
+    "@amplience/gql-ai-sdk": "1.1.2--canary.14.4b4a614.0",
```
Can we use latest?

amplience-dling commented on src/stories/AmplienceImageStudio.stories.ts:
```diff
@@ -0,0 +1,149 @@
+      'https://thumbs.amplience-qa.net/r/a124da68-5b5d-46a7-94dc-13a4c45976f8',
+    imageName: 'TestImageName',
+    options: {
+      baseUrl: 'https://qa-mandelbrot-image-studio.amplience-qa.net/image-studio/',
```
Should this be this url? if no, its used in this file a few times

bennealon replied:
qa-release was pointing to the older topic/shoptalk branch, and therefore was out of sync with the messaging contract that this SDK uses. Now that QA-Release is aligned with latest, this has been updated.

bennealon committed:
Update src/types/ImageStudio.types.ts

Co-authored-by: rs-amp <56435292+rs-amp@users.noreply.github.com>

bennealon committed:
Update src/types/ImageStudio.types.ts

Co-authored-by: rs-amp <56435292+rs-amp@users.noreply.github.com>

bennealon committed:
Addition of `editImages` and `launch` SDK actions
removes caching of SDK promise
Changes to SDK types to faciliate allowing multiple images to be passed to image-studio. Despite the SDK passing multiple images to the studio, only the first image is ever loaded.
Updates Storybook to use local fixture
Fixes issue with image being too large for image-studio (>16MP)
Fixes some bad formatting

bennealon committed:
build updates to access github packages on public repos

bennealon committed:
fix: commonJS and ESM should be compatible respectively once built
fix: package defaults to commonjs, but builds to es2022 on --module cli command
chore: updates gql-ai-sdk to
latest release

bennealon committed:
allows SDK to control imageSave and logout behaviour in image-studio

bennealon committed:
updates allow `launch` and `editImages` contracts to modify the logout and saveImage behaviour of imageStudio

bennealon committed:
target es6/es2015 for this SDK

bennealon committed:
fix: removes dependency on internal amplience library, image-studio will itnernally convert non-whitelisted domain url's on ingestion

Benj0s commented:
Tested OK on extension integrations - https://github.com/amplience/dc-extension-shoppable-image & https://github.com/amplience/dc-extension-di-transform

bennealon committed:
Updates SDKImage to transmit/receive the image mimeType

bennealon committed:
removes redundant amplience registry npmrc
moves prettier into its own file

bennealon committed:
removes QA and localhost ref from stories

rs-amp commented on src/types/Message.ts:
```diff
@@ -0,0 +1,27 @@
+ */
+export interface SDKMetadata {
+  allowImageSave: boolean;
+  allowLogout: boolean;
```
I suppose this means that if you provide a metadata object, you need to provide _both_ values? (not providing the object makes it use default values, I'm guessing)

bennealon replied:
No, this object is never user-supplied and intentionally explicit, the options are handled (for now) internally within the `AmplienceImageStudio` class - but we could expose this in future. ImageStudio itself has been updated to look specifically for these keys https://github.com/amplience/image-studio/pull/274

rs-amp commented on src/types/ImageStudio.ts:
```diff
@@ -0,0 +1,24 @@
+}
+
+/**
+ * container interface for image representation
```
```suggestion
 * Container interface for image representation
```
Capitalizing this would make it consistent with other docs, since they'll be visible to users of the library.

rs-amp commented on src/AmplienceImageStudio.ts:
```diff
@@ -0,0 +1,175 @@
+  constructor(protected options: AmplienceImageStudioOptions) {}
+
+  /**
+   * launches image studio to edit the images
```
Might want to use proper casing and punctuation for these docs, since they'll pop up when people are using the library.

rs-amp reviewed (commented):
Looks good, just a few comments about docs styling.

The storybook interaction tests are a bit weird for this right now. When you browse to them, it instantly opens image studio without any explanation, and you have to manually close it to get the test result. It would probably be better if image studio's launch were somehow mocked so that these tests could be automated (and potentially run on CI).

bennealon committed:
Update src/types/ImageStudio.ts

Co-authored-by: rs-amp <56435292+rs-amp@users.noreply.github.com>

bennealon commented:
> Looks good, just a few comments about docs styling.
> 
> The storybook interaction tests are a bit weird for this right now. When you browse to them, it instantly opens image studio without any explanation, and you have to manually close it to get the test result. It would probably be better if image studio's launch were somehow mocked so that these tests could be automated (and potentially run on CI).

Currently, its by design to launch image-studio, the story title describes whats expected for the user to do, and once the studio closes the expectations are checked. There are disadvantages to using a live service as part of these tests, one of them is whats been highlighted here. We could update in future to include a button to begin the test.

These tests are currently an interactive showcase of how to use the SDK, they assert the correct behaviour as a by-product. I agree, this isn't useful as a CI test scenario right now and we can look at addressing this in future. Theres not alot to test without interacting with a live service, so i feel if we mocked the required behaviour there would be little value in the tests.

rs-amp commented:
> We could update in future to include a button to begin the test.

I feel like creating these scenarios is what the "try me" story in the storybook is for, though you do need to manually select the parameters each time.
